### PR TITLE
libyottadb Simple API user documentation draft

### PR DESCRIPTION
**Pull request by ksbhaskar**

### CODE REVIEW

**Comment by ChristopherEdwards** [comment]
I don't have full context and did a quick breeze through the documentation:

Normally you don't commit binary files (pdf) to source control, based on what i see in the commits the *.rst documents contain all of the information contained within the pdf. You can add binaries for the PDF documentation when you do a release: https://help.github.com/articles/creating-releases/

Also, note that the pdf web render also appears to be broken in github so none of the right hand side of the equals sign values appear correctly (at least on my machine).

**Comment by ksbhaskar** [comment]
Yes, you're right. I should take the PDF out of Github (the PDF is
generated from the rst using rst2pdf). Thanks.

-- Bhaskar

**Comment by shabiel** [comment]
I don't see any APIs to call existing M code. Something like ydb_function and ydb_procedure.

**Comment by ksbhaskar** [comment]
There is no need for a new API to call existing M code - they already exist
in YottaDB, and the Simple API takes nothing away from YottaDB. Were you
looking for something different?

-- Bhaskar

**Comment by shabiel** [comment]
Here's a real life example that I had to write yesterday:

https://github.com/shabiel/ewd-vista-pharmacy/blob/master/lib/ewd-vista-pharmacy.js#L603

**Comment by ksbhaskar** [comment]
Thanks for your comments, Sam.

There is an important difference between calling from C to M vs. calling C
to C (the libyottadb simple API). When calling from C to M:


   - As the namespace and syntax of
   ​called ​
   M entryrefs are different from those of C function
   ​ entry points, there needs to be a mapping from one to the other.
   - As YottaDB does the type conversion, there is a need to specify what
   the types are of the C data to be converted.


When calling from C to C:


   - The names and syntax of the called routines are all C function entry
   points; there is no need for a mapping.
   - As the caller is responsible for all type conversion - all data is
   passed as strings - there is no need to specify in the interface what
   conversion is required.


Therefore, while calling C to M requires a call-in table, calling C to C
does not. So, while an application would use the existing call-in to call
Fileman and other APIs exposed by VistA, and needs a call-in table to do
so, it can call the libyottadb Simple API directly without the call-in
table (the "call-in table" in effect is the .h file that is #include'd at
compile time).

Passing by reference is an interesting question. In effect, it is passing a
local variable structure from a caller to a called routine. There is no C
structure equivalent to a local variable, and the only way to pass a
structure is to pass a pointer. So, as a practical matter:


   - The only way to maintain a complete local variable structure in a C
   application is to maintain it within YottaDB and access
   ​node ​
   values.
   - The only way to effect passing a structure is to pass a name as a
   string using the C to M interface, and have an M shim function that passes
   the name to the actual function. We'll think about the ramifications of an
   enhancement in the C to M function that passes a string which is to be
   treated as an M variable name.


All the capabilities of $order() are exposed by the ydb_subscript_next()
and ydb_subscript_previous() functions. The Simple API as currently drafted
does not include support for M locks or indirection - let us know if those
are required.

Regards
-- Bhaskar

**Comment by shabiel** [comment]
Thank you for the detailed reply.

Yes, locks are needed in VistA. We can't quite use transactions.

No, I don't think I need to use indirection.

--Sam

File: libyottadb/SimpleAPI.rst [hunk 1]
```diff
@@ -0,0 +1,1010 @@
...
+**Caveat:** This code does not exist yet. The user documentation is
+being written ahead of the code, and will change in the event the code
+needs to differ from this document for a valid technical reason. Also,
+this document itself is incomplete and still evolving.
+
+===========
+Quick Start
+===========
+
+**The Quick Start section needs to be fleshed out.**
```
Comment (nars1) [inline libyottadb/SimpleAPI.rst]: Is this still relevant?

File: libyottadb/SimpleAPI.rst [hunk 2]
```diff
@@ -0,0 +1,1010 @@
...
+  the called function returns a ``YDB_TP_RESTART`` return code, it
+  calls the function again.
+- If the function returns a ``YDB_TP_ROLLBACK``, `ydb_tp_s()`_ returns
+  to its caller with that return code.
+- To protect applications against poorly coded transactions, if a
+  transaction takes longer than the number of seconds specified by
+  the environment variable ``ydb_maxtptime``, libyottadb aborts the
+  transaction and the `ydb_tp_s()`_ function returns the
+  ``YDB_ERR_TPTIMEOUT`` error.
+
```
Comment (nars1) [inline libyottadb/SimpleAPI.rst]: If the transaction is stuck in C code (as opposed to M code), then we might not be able to abort it until the next M statement is reached. Does that need mention here?

**Pull request merged**